ansible_events_ui/db/utils/lostream.py

```diff
@@ -0,0 +1,247 @@
+import logging
+import os
+from io import SEEK_SET, UnsupportedOperation
+from typing import Tuple, Union
+
+from fastapi import Depends
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from ansible_events_ui.db.dependency import get_db_session
+
+LOG = logging.getLogger(__name__)
+
+
+# Default read buffer size is
+CHUNK_SIZE = 1024**2
+
+
+# Class that will use server-side functions to access
+# PostgreSQL large object
+class LObject:
+    """
+    Facilitate PostgreSQL large object functionality using the
+    PostgreSQL server-side functions.
+
+    As of 2022-09-01, there is no large object support directly
+    in asyncpg or other async Python PostgreSQL drivers.
+    """
+
+    VALID_MODES = {"rwt", "rt", "wt", "at", "rwb", "rb", "wb", "ab"}
+    INV_READ = 0x00040000
+    INV_WRITE = 0x00020000
+    MODE_MAP = {"r": INV_READ, "w": INV_WRITE, "a": INV_READ | INV_WRITE}
+
+    def __init__(
+        self: "LObject",
+        oid: int = 0,
+        length: int = 0,
+        session: AsyncSession = None,
+        *,
+        chunk_size: int = CHUNK_SIZE,
+        mode: str = "rb",
+    ):
+        LOG.debug(
+            f"LObject Init with oid={oid}, length={length}, chunk_size={chunk_size}"
+        )
+        self.session = session
+        self.chunk_size = chunk_size if chunk_size > 0 else CHUNK_SIZE
+        self.oid = oid
+        self.length = length
+        self.closed = False
+
+        self.imode, self.text_data, self.append = self._resolve_mode(mode)
+        self.pos = self.length if self.append else 0
+
+    def _resolve_mode(self: "LObject", mode: str):
+        if mode not in self.VALID_MODES:
+            raise ValueError(
+                f"Mode {mode} must be one of {sorted(self.VALID_MODES)}"
+            )
+
+        _text_data = mode.endswith("t")
+        _append = mode.startswith("a")
+        _mode = 0
+        for c in mode:
+            _mode |= self.MODE_MAP.get(c, 0)
+
+        LOG.debug(
+            f"LObject mode '{mode}' resolved to _mode={_mode}, _text_data={_text_data}, _append={_append}"
+        )
+
+        return _mode, _text_data, _append
+
+    def closed_check(self: "LObject"):
+        if self.closed:
+            raise UnsupportedOperation("Large object is closed")
+
+    async def __aenter__(self) -> "LObject":
```

Comment:
You use this string in many places. I suggest do instead:
```
L = typing.TypeVar("L", bound="LObject")

[...]
async def __aenter__(self) -> L:
```
Which is also compatible with subclassing. Also you can do the following to be able to refer to itself. 
```
from __future__ import annotations
[...]
async def __aenter__(self) -> LObject:
```
Honestly, I don't know what of both would be better. Maybe @cutwater knows.

Reply:
If we go ahead and upgrade to 3.10, I can drop the string references and just use `LObject` directly.
I like it spelled out. Abbreviations just mean you have to figure out something else.

Reply:
The import `from __future__ import annotations` works for 3.9 as well and in my opinion it's a preferrable way of forward reference definition before 3.10.
I don't think we can upgrade to 3.10 at the moment, since it's not supported by RHEL and there is no official container image with python 3.10 installed.
The TypeVar, that @Alex-Izquierdo proposed is a workaround for `Self` type, which is introduced in 3.11, but I'll leave the decision to your discretion. I don't think this class is expected to be inherited, so introducing a TypeVar isn't that necessary, but rather nice to have.

Reply:
Using  __future__